RoboJackets/robocup-software
# Planner ball obstacle

## Description

This PR updates the planning system to consider the ball a dynamic obstacle, and plan paths that keep a certain distance away from the ball as published by the coach node

## Associated / Resolved Issue
Resolves [make planner respect ball on stop](https://app.clickup.com/t/34c4duz)

## Design Documents
[Link](link-to-design-doc)

## Steps to Test
### Test Case 1
1. Run the sim (green check/normal mode). Currently, six robots move in a predefined manner. Robot 1 (the defender) is the only one set to pay any attention to the ball. 
2. Place the ball in/near the path of robot 1
3. Give the ball some velocity in a collision course with robot 1's path

**Expected result:** Robot 1 avoids touching the ball

### Test Case 2: STOP
1. Run the sim in STOP. (Coach node should be publishing a min_dist_from_ball at 0.5)
2. Place the ball in/near the path of robot 1
4. Give the ball some velocity in a collision course with robot 1's path

**Expected result:** Robot 1 should keep a larger radius (0.5 in some unit?) around the ball.


## Key Files to Review
Plan Request: struct changed to keep a `min_dist_from_ball` field. also changed `fill_obstacles` (utility called by planners) to use the field.
 * `plan_request.hpp`
 * `plan_request.cpp`

Planner Node: now maintains a `coach_state` subscription and inserts the `min_dist_from_ball` data into each plan request when creating them from RobotIntents.
 * `planner_node.hpp`
 * `planner_node.cpp`

## Review Checklist
